plugin/parser/tokenizer.go

```diff
@@ -0,0 +1,284 @@
+// We mainly considered:
+//   - comments
+//     - style /* comments */
+//     - style -- comments
+//   - string
+//     - style 'string'
+//     - style $$ string $$
+//   - identifier
+//     - style "indentifier"
+//
+// Notice:
+//   - We support PostgreSQL CREATE PROCEDURE statement with $$ $$ style,
+//       but do not support BEGIN ATOMIC ... END; style.
+//       See https://www.postgresql.org/docs/14/sql-createprocedure.html.
+func (t *tokenizer) splitPostgreSQLMultiSQL() ([]string, error) {
+	var res []string
+
+	t.skipBlank()
+	startPos := t.cursor
+	for {
+		switch {
+		case t.char(0) == '/' && t.char(1) == '*':
+			if err := t.scanComment(); err != nil {
+				return nil, err
+			}
+		case t.char(0) == '-' && t.char(1) == '-':
+			if err := t.scanComment(); err != nil {
+				return nil, err
+			}
+		case t.char(0) == '\'':
+			if err := t.scanString('\''); err != nil {
+				return nil, err
+			}
+		case t.char(0) == '$':
+			if err := t.scanDollarQuotedString(); err != nil {
+				return nil, err
+			}
+		case t.char(0) == '"':
+			if err := t.scanIdentifier('"'); err != nil {
+				return nil, err
+			}
+		case t.char(0) == ';':
+			t.skip(1)
+			res = append(res, t.getString(startPos, t.pos()-startPos))
+			t.skipBlank()
+			startPos = t.pos()
+		case t.char(0) == eofRune:
+			s := t.getString(startPos, t.pos())
+			if !emptyString(s) {
+				res = append(res, s)
+			}
+			return res, nil
+		default:
+			t.skip(1)
+		}
+	}
+}
+
+// Assume that identifier only contains letters, underscores, digits (0-9), or dollar signs ($).
+// See https://www.postgresql.org/docs/current/sql-syntax-lexical.html.
+func (t *tokenizer) scanIdentifier(delimiter rune) error {
+	if t.char(0) != delimiter {
+		return fmt.Errorf("delimiter doesn't start with delimiter: %c, but found: %c", delimiter, t.char(0))
+	}
+
+	t.skip(1)
+	for {
+		switch t.char(0) {
+		case delimiter:
+			t.skip(1)
+			return nil
+		case eofRune:
+			return fmt.Errorf("invalid indentifier: not found delimiter: %c, but found EOF", delimiter)
+		default:
+			t.skip(1)
+		}
+	}
+}
+
+// There are two ways to include a single quote('), using \' or ''.
+// We only handle the case \', because the second case '' does not require special handling.
+// In more detail, we can think of
+//     'this is a string contains ''.'
+// as
+//     'this is a string contains '
+// and
+//     '.'
+//.
+// And this is extensible.
+func (t *tokenizer) scanString(delimiter rune) error {
+	if t.char(0) != delimiter {
+		return fmt.Errorf("string doesn't start with delimiter: %c, but found: %c", delimiter, t.char(0))
+	}
+
+	t.skip(1)
+	for {
+		switch t.char(0) {
+		case delimiter:
+			t.skip(1)
+			return nil
+		case eofRune:
+			return fmt.Errorf("invalid string: not found delimiter: %c, but found EOF", delimiter)
+		case '\\':
+			// skip two because we want to skip \' and \\.
+			t.skip(2)
+		default:
+			t.skip(1)
+		}
+	}
+}
+
+func (t *tokenizer) scanDollarQuotedString() error {
```

Comment:
nit: should this be called scanDoubleDollarQuotedString? If that's the case, line 64 should be case t.char(0) == '$' && case t.char(1) == '$'.

If this is used for both single and double-dollar delimiters, I wonder if we could use separate functions to avoid the KMP algorithms below.

Reply:
Sorry for the confusion. I rename it as `scanDoubleDollarQuotedString`.

But I think line 64 cannot be `case t.char(0) == '$' && case t.char(1) == '$'`, because there are two double dollar quoted string syntax:
- $$ string $$
- $tag$ string $tag$

Reply:
reference: https://www.postgresql.org/docs/current/sql-syntax-lexical.html

Reply:
Per offline chat, let's add some comments.